### PR TITLE
Experimental subplot implementation that mimics matplotlib's style

### LINKED ISSUE #20: Implement subplot command
This is a very cool new command.

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Experimenting with a more matplotlib-like syntax for managing subplots (an offshoot from #412). Preview of tutorial is at https://pygmt-git-fancysubplots.gmt.now.sh/tutorials/subplots.html, which is actually translated and adapted from the great example at https://docs.gmt-china.org/latest/tutorial/subplot/ (in Chinese). Specifically, this mimics the [`fig, axs = plt.subplots()` style](https://stackoverflow.com/a/55771757/6611055) of defining subplots! An example of how it looks like currently:

```python
import pygmt

fig, axs = pygmt.subplots(nrows=1, ncols=2, figsize=("6c", "3c"))
fig.sca(ax=axs[0, 0])
fig.basemap(region=[0, 3, 0, 3], frame=True)
fig.sca(ax=axs[0, 1])
fig.basemap(region=[0, 3, 0, 3], frame=True)
fig.end_subplot()
fig.show()

# or

fig, axs = pygmt.subplots(nrows=1, ncols=2, figsize=("6c", "3c"))
fig.basemap(region=[0, 3, 0, 3], frame=True, ax=axs[0, 0])
fig.basemap(region=[0, 3, 0, 3], frame=True, ax=axs[0, 1])
fig.end_subplot()
fig.show()
```

produces:

Basic Subplot

The code isn't perfect yet, and I'm very open to suggestions on how to go about improving this implementation. Specifically, it would be cool if we could do `ax.plot()` like in matplotlib. Calling expert Pythonistas for tips :wave:

As it currently stands, there's a high-level `subplots` function in subplot.py (somewhat based on the matplotlib `plt.subplots` equivalent at https://github.com/matplotlib/matplotlib/blob/v3.2.1/lib/matplotlib/pyplot.py#L1044). This wraps around the new class SubPlot (inherited from the main Figure class to include subplot functionality). Note that since `set` is a Python reserved character so I've used `sca` instead to set the current active subplot, following matplotlib.

Also cross-referencing Julia wrapper/implementation at https://www.generic-mapping-tools.org/GMT.jl/latest/#GMT.subplot.

Parameters/Aliases to wrap:

- [x] F = dims/dimensions/size/sizes
- [x] A = autolabel
- [x] B = frame
- [x] C = clearance
- [ ] J = projection
- [x] M = margins
- [x] R = region
- [x] S (C/R) = layout
- [x] T= title

References:
- https://matplotlib.org/3.2.1/gallery/subplots_axes_and_figures/subplots_demo.html
- https://matplotlib.org/3.2.1/tutorials/introductory/pyplot.html
- https://stackoverflow.com/a/55771757/6611055
- https://matplotlib.org/api/_as_gen/matplotlib.pyplot.sca.html
- https://nbviewer.jupyter.org/github/SHTOOLS/SHTOOLS/blob/master/examples/notebooks/plotting-maps.ipynb#Creating-Subplots
- https://basemaptutorial.readthedocs.io/en/latest/subplots.html

TODO:
- [x] Initial implementation
- [x] Alias `ax` to the [`-c`](https://docs.generic-mapping-tools.org/latest/gmt.html#c-full) command in GMT plotting functions
- [ ] Find a way to call `ax[0, 1].plot()` directly instead of using `fig.sca(ax)`.

Addresses https://github.com/GenericMappingTools/pygmt/issues/366#issuecomment-549386681, Supersedes and Closes #412, Fixes #20.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

### CODE REVIEW

**Comment by leouieda** [comment]
Hi Wei Ji, thanks for doing this! I’ve been looking through this a bit now. I like the subplots function! But I think we shouldn’t do sca as well. A major fault of matplotlib is having too many ways of doing the same thing. We can stick close to their api but need to be careful not to repeat their mistakes.

**Comment by weiji14** [comment]
> But I think we shouldn’t do sca as well. A major fault of matplotlib is having too many ways of doing the same thing. We can stick close to their api but need to be careful not to repeat their mistakes.

Yes, the good ol [Zen of Python](https://www.python.org/dev/peps/pep-0020/) - "There should be one-- and preferably only one --obvious way to do it.". I definitely prefer using `ax=` to set the current subplot.

However, the `sca` function (which is basically an alias for GMT subplot's [`set`](https://docs.generic-mapping-tools.org/latest/subplot.html#synopsis-set-mode)) would be helpful say, if we want to label our subplots "abcd" or set a clearance/space between subplots. Would there be a better way to do those without resorting to using `sca`?

**Comment by weiji14** [comment]
Ping @MarkWieczorek, I noticed that SHTOOLS has a tutorial on creating subplots using PyGMT's using X and Y offsets at https://nbviewer.jupyter.org/github/SHTOOLS/SHTOOLS/blob/master/examples/notebooks/plotting-maps.ipynb#Creating-Subplots. Right now this PR is attempting to have a syntax like `fig, ax = pygmt.subplots(nrows=1, ncols=2)` which could make it easier to create subplots in the future (i.e. without relying on X and Y offsets).

Would be great to get hear your thoughts on this, e.g. would you like to do `grid.plotgmt(..., ax=ax[0])` in SHTOOLS? Let us know if there's anything in particular from `matplotlib` you would like or not like to have implemented here :grin:

**Comment by MarkWieczorek** [comment]
This looks good to me! I would be happy to test this with shtools, but this will have to wait until mid october....

File: examples/tutorials/subplots.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+###############################################################################
+# For example, to activate the subplot on the top right corner (index: 2) at
+# ``row=0`` and ``col=2``, so that all subsequent plotting commands happen
+# there, you can use the following command:
+
+###############################################################################
+# .. code-block:: default
+#
+#     fig.sca(ax=axs[0, 2])
```
Comment (weiji14) [inline examples/tutorials/subplots.py:77]: Note to self to remove mention of `fig.sca`, if we're going to hide this away as per https://github.com/GenericMappingTools/pygmt/pull/427#issuecomment-646983555.

**Review by weiji14 — COMMENTED** [review]
Alright, I'm marking this as ready to review, though there are still some rough spots. The translation over from https://docs.gmt-china.org/latest/tutorial/subplot/ is complete, and I think it's worth the team taking a look. Let me know if I should cherry-pick stuff out into separate PRs to make it easier to review.

Note that there are some functionality from https://docs.generic-mapping-tools.org/latest/subplot.html that aren't covered yet:
- `-Fs` (subfigsize?) at https://docs.generic-mapping-tools.org/latest/subplot.html#f
- The matplotlib way of plotting using something like `ax[0, 1].plot()` - quite tricky to implement!!

**Pull request closed without merging**